#
# This is the user-interface definition of a Shiny web application. You can
# run the application by clicking 'Run App' above.
#
# Find out more about building applications with Shiny here:
# 
#    http://shiny.rstudio.com/
#

library(shiny)

# Define UI for application that draws a histogram
shinyUI(fluidPage(
  
  # Application title
  titlePanel("Seurat Shiny"),
  fluidRow(
    uiOutput('debug_ui'),
    uiOutput('select_dataset_ui'),
    uiOutput('reduce_dataset_select_ui'),
    actionButton('load_data','Load_Data'),
    #selectInput('select_dataset','Select Dataset',datasets,datasets[1]),
    textOutput('data_path_text'),
    textOutput('dataset_text'),
    tabsetPanel(selected = 'Data Processing',
      tabPanel('Upload Data',
               tags$h5('Upload a tar.gz file containing the outs folder generated by CellRanger, the name of the containing folder will be used to name the dataset'),
               fileInput("file", "Upload Zip file", accept = ".gz"),
               actionButton("unzip", "Unzip Files")
            
               ),
      tabPanel('Data Processing',
               tabsetPanel(selected = '',
      tabPanel('Filtering and Normalisation',
               textOutput('test_test'),
               
               tags$h4('### Standard pre-processing workflow'),
               tags$h5('
                       The steps below encompass the standard pre-processing workflow for scRNA-seq data in Seurat. These represent the creation of a Seurat object, the selection and filtration of cells based on QC metrics, data normalization and scaling, and the detection of highly variable genes. In previous versions, we grouped many of these steps together in the `Setup` function, but in v2, we separate these steps into a clear and sequential workflow.
                       '),
                       
                tags$h4('### QC and selecting cells for further analysis'),
                tabsetPanel(
                  tabPanel('Create Seurat Object',
                  tags$h5('While the `CreateSeuratObject` imposes a basic minimum gene-cutoff, you may want to filter out cells at this stage based on technical or biological parameters. Seurat allows you to easily explore QC metrics and filter cells based on any user-defined criteria. In the example below, we visualize gene and molecule counts, plot their relationship, and exclude cells with a clear outlier number of genes detected as potential multiplets. Of course this is not a guaranteed method to exclude cell doublets, but we include this as an example of filtering user-defined outlier cells. We also filter cells based on the percentage of mitochondrial genes present.'),
                  column(6,numericInput('min.cells','Minimum number of cells per gene',step = 1,min = 1,value = 3)),
                  column(6,numericInput('min.genes','Minimum number of genes per cell',step = 1,min = 1,value = 200)),
                  
                  uiOutput('remove_select_gene_ui'),
                  plotOutput('remove_hist'),
                  numericInput('remove_threshold','Select Expression Threshold',1),
                  htmlOutput('text_remaining'),
                  actionButton('run_reduce','Save Data'),
                  
                  column(6,actionButton('reload_data','Reload Data')),
                  column(12,plotOutput('features_plot'))
                ),
               tabPanel("Gene Plot",
                tags$h5('After removing unwanted cells from the dataset, the next step is to normalize the data. By default, we employ a global-scaling normalization method "LogNormalize" that normalizes the gene expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result. '),
                plotOutput('gene_plot')
               ),
              tabPanel('Filter, Normalise and Find Variable Genes',
                tags$h5('Seurat calculates highly variable genes and focuses on these for downstream analysis. **`FindVariableGenes`** calculates the average expression and dispersion for each gene, places these genes into bins, and then calculates a z-score for dispersion within each bin. This helps control for the relationship between variability and average expression. This function is unchanged from (Macosko *et al*.), but new methods for variable gene expression identification are coming soon. We suggest that users set these parameters to mark visual outliers on the dispersion plot, but the exact parameter settings may vary based on the data type, heterogeneity in the sample, and normalization strategy. The parameters here identify ~2,000 variable genes, and represent typical parameter settings for UMI data that is normalized to a total of 1e4 molecules.'),
                textOutput('FilterCells_text'),
                textOutput('NormalizeData_text'),
                textOutput('FindVariableGenes_text'),
                plotOutput('var_plot')
               ))
               ),
      tabPanel('Scaling',
               
               tags$h4('### Scaling the data and removing unwanted sources of variation'),
               
               tags$h5("Your single cell dataset likely contains 'uninteresting' sources of variation. This could include not only technical noise, but batch effects, or even biological sources of variation (cell cycle stage). As suggested in [Buettner *et al*, NBT, 2015](https://www.nature.com/articles/nbt.3102), regressing these signals out of the analysis can improve downstream dimensionality reduction and clustering. To mitigate the effect of these signals, Seurat constructs linear models to predict gene expression based on user-defined variables. The scaled z-scored residuals of these models are stored in the scale.data slot, and  are used for dimensionality reduction and clustering."),
               
               tags$h5("We can regress out cell-cell variation in gene expression driven by batch (if applicable), cell alignment rate (as provided by Drop-seq tools for Drop-seq data),  the number of detected molecules, and mitochondrial gene expression. For cycling cells, we can also learn a 'cell-cycle' score (see example [here](http://satijalab.org/seurat/cell_cycle_vignette.html)) and regress this out as well. In this simple example here for post-mitotic blood cells, we regress on the number of detected molecules per cell as well as the percentage mitochondrial gene content. "), 
               
               tags$h5("Seurat v2.0 implements this regression as part of the data scaling process. Therefore, the `RegressOut` function has been deprecated, and replaced with the vars.to.regress argument in `ScaleData`."),
               
               actionButton('run_scaling','Run Scaling'),
               textOutput('ScaleData_text'),
               textOutput('scaling_done_text')
               
                ),
      
      tabPanel('Remove Data'
               
               ),
      tabPanel('PCA',
               tags$h4('### Perform linear dimensional reduction'),
               
               tags$h5("Next we perform PCA on the scaled data. By default, the genes in `object@var.genes` are used as input, but can be defined using pc.genes. We have typically found that running dimensionality reduction on highly variable genes can improve performance. However, with UMI data - particularly after regressing out technical variables, we often see that PCA returns similar (albeit slower) results when run on much larger subsets of genes, including the whole transcriptome."),
               radioButtons('pca_full_dataset','Full Dataset',c(T,F)),
               actionButton('run_pca','Run PCA'),
               
               textOutput('pca_1'),
               verbatimTextOutput('pca_text'),
               plotOutput('VizPCA'),
               plotOutput('VizPlot')
               
               ),
      tabPanel('PCAHeatmap',
               tags$h5("In particular `PCHeatmap` allows for easy exploration of the primary sources of heterogeneity in a dataset, and can be useful when trying to decide which PCs to include for further downstream analyses. Both cells and genes are ordered according to their PCA scores. Setting cells.use to a number plots the 'extreme' cells on both ends of the spectrum, which dramatically speeds plotting for large datasets. Though clearly a supervised analysis, we find this to be a valuable tool for exploring correlated gene sets."),
               actionButton('ProjectPCA','Run ProjectPCA'),
               plotOutput('PCHeatmap'),
               plotOutput('PCHeatmap_2')
               
               ),
      tabPanel('PCA Statistics',
               tags$h4("### Determine statistically significant principal components"),

               tags$h5("To overcome the extensive technical noise in any single gene for scRNA-seq data, Seurat clusters cells based on their PCA scores, with each PC essentially representing a 'metagene' that combines information across a correlated gene set. Determining how many PCs to include downstream is therefore an important step."),
                       
              tags$h5("In [Macosko *et al*](http://www.cell.com/abstract/S0092-8674(15)00549-8), we implemented a resampling test inspired by the jackStraw procedure. We randomly permute a subset of the data (1% by default) and rerun PCA, constructing a 'null distribution' of gene scores, and repeat this procedure. We identify 'significant' PCs as those who have a strong enrichment of low p-value genes."),
              actionButton('Jacksaw', 'Run Jacksaw'),
              
              tags$h5("The `JackStrawPlot` function provides a visualization tool for comparing the distribution of p-values for each PC with a uniform distribution (dashed line). 'Significant' PCs will show a strong enrichment of genes with low p-values (solid curve above the dashed line). In this case it appears that PCs 1-10 are significant."),
              plotOutput('JackStrawPlot'),
              
              tags$h5("A more ad hoc method for determining which PCs to use is to look at a plot of the standard deviations of the principle components and draw your cutoff where there is a clear elbow in the graph. This can be done with `PCElbowPlot`. In this example, it looks like the elbow would fall around PC 9."),
              plotOutput('PCElbowPlot'),
              
              tags$h5("PC selection -- identifying the true dimensionality of a dataset -- is an important step for Seurat, but can be challenging/uncertain for the user. We therefore suggest these three approaches to consider. The first is more supervised, exploring PCs to determine relevant sources of heterogeneity, and could be used in conjunction with GSEA for example. The second implements a statistical test based on a random null model, but is time-consuming for large datasets, and may not return a clear PC cutoff. The third is a heuristic that is commonly used, and can be calculated instantly. In this example, all three approaches yielded similar results, but we might have been justified in choosing anything between PC 7-10 as a cutoff. We followed the jackStraw  here, admittedly buoyed by seeing the PCHeatmap returning interpretable signals (including canonical dendritic cell markers) throughout these PCs. Though the results are only subtly affected by small shifts in this cutoff (you can test below), we strongly suggest always explore the PCs they choose to include downstream.")
              
               ),
      tabPanel('Cluster Cells',
               tags$h4("### Cluster the cells"),
               
               tags$h5("Seurat now includes an graph-based clustering approach compared to (Macosko *et al*.). Importantly, the *distance metric* which drives the clustering analysis (based on previously identified PCs) remains the same. However, our approach to partioning the cellular distance matrix into clusters has dramatically improved. Our approach was heavily inspired by recent manuscripts which applied graph-based clustering approaches to scRNA-seq data [[SNN-Cliq, Xu and Su, Bioinformatics, 2015]](http://bioinformatics.oxfordjournals.org/content/early/2015/02/10/bioinformatics.btv088.abstract) and CyTOF data [[PhenoGraph, Levine *et al*., Cell, 2015]](http://www.ncbi.nlm.nih.gov/pubmed/26095251). Briefly, these methods embed cells in a graph structure - for example a K-nearest neighbor (KNN) graph, with edges drawn between cells with similar gene expression patterns, and then attempt to partition this graph into highly interconnected 'quasi-cliques' or 'communities'. As in PhenoGraph, we first construct a KNN graph based on the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity). To cluster the cells, we apply modularity optimization techniques such as the Louvain algorithm (default) or SLM [[SLM, Blondel *et al*., Journal of Statistical Mechanics]](http://dx.doi.org/10.1088/1742-5468/2008/10/P10008), to iteratively group cells together, with the goal of optimizing the standard modularity function."),
               
               tags$h5("The `FindClusters` function implements the procedure, and contains a resolution parameter that sets the 'granularity' of the downstream clustering, with increased values leading to a greater number of clusters. We find that setting this parameter between 0.6-1.2 typically returns good results for single cell datasets of around 3K cells. Optimal resolution often increases for larger datasets. The clusters are saved in the `object@ident` slot."),
               actionButton('find_clusters','Run FindClusters')
               ),
      tabPanel('tSNE',
               tags$h4("### Run Non-linear dimensional reduction (tSNE)"),

               tags$h5("Seurat continues to use tSNE as a powerful tool to visualize and explore these datasets. While we no longer advise clustering directly on tSNE components, cells within the graph-based clusters determined above should co-localize on the tSNE plot. This is because the tSNE aims to place cells with similar local neighborhoods in high-dimensional space together in low-dimensional space. As input to the tSNE, we suggest using the same PCs as input to the clustering analysis, although computing the tSNE based on scaled gene expression is also supported using the genes.use argument."),
               actionButton('run_tsne','Run TSNE'),
               plotOutput('TSNEPlot')
      ))
               ),
      tabPanel('Visualisation',
               textOutput('values_text'),
               tabsetPanel(
      tabPanel('Differential Genes',
               tags$h4("### Finding differentially expressed genes (cluster biomarkers)"),
               
               tags$h5("Seurat can help you find markers that define clusters via differential expression. By default, it identifes positive and negative markers of a single cluster (specified in `ident.1`), compared to all other cells.  `FindAllMarkers` automates this process for all clusters, but you can also test groups of clusters vs. each other, or against all cells."),
               
               tags$h5("The `min.pct` argument requires a gene to be detected at a minimum percentage in either of the two groups of cells, and the thresh.test argument requires a gene to be differentially expressed (on average) by some amount between the two groups. You can set both of these to 0, but with a dramatic increase in time - since this will test a large number of genes that are unlikely to be highly discriminatory. As another option to speed up these computations, `max.cells.per.ident` can be set. This will downsample each identity class to have no more cells than whatever this is set to. While there is generally going to be a loss in power, the speed increases can be significiant and the most highly differentially expressed genes will likely still rise to the top."),
               column(12,selectInput('find_single_cluster_test',"Test to Use",FindMarkers_test_list,selected = "Wilcoxon rank sum test (default)")),
               column(4,numericInput('logfc.threshold','logfc.threshold',value = 0.25)),
               column(4,numericInput('min.pct','min.pct',value = 0.1)),
               column(4),
               column(12,
               tabsetPanel(
                 tabPanel('Top markers for All Clusters',
                          
                          actionButton('find_all_clusters','Run All Cluster'),
                          tags$h4('This can take up to half and hour to process'),
                          numericInput('clust_num_display','Number of Markers to Display',value = 2),
                          dataTableOutput('data_markers_table')
                 ),
                 tabPanel('Single Cluster',
                    column(2,numericInput('single_cluster_number','Cluster Number',min = 0, max = 50,step = 1,value = 1)),
                    column(10),
                  column(12,
                    tabsetPanel(
                      tabPanel('Single',
                        column(3,actionButton('find_single_cluster','Run Single Cluster FindMarker')),
                    
                        column(12,dataTableOutput('single_markers_table'))
                    
                 ),
                 tabPanel('Gene Unique to Cluster',
                    actionButton('find_distinguishing_cluster','Run Distinguish Cluster '),
                    selectInput('find_distinguishing_cluster_select','Select Clusters',c(1:50),selected = c(2,3),multiple = T),
                    dataTableOutput('distinguish_markers_table')
                 ))))
))
     
        
               ),
      tabPanel('Visualise Marker Expression',
               tags$h5("We include several tools for visualizing marker expression. `VlnPlot` (shows expression probability distributions across clusters), and `FeaturePlot` (visualizes gene expression on a tSNE or PCA plot) are our most commonly used visualizations. We also suggest exploring `RidgePlot`, `CellPlot`, and `DotPlot` as additional methods to view your dataset."),
               
               
               uiOutput('plot_select_genes_ui'),
               plotOutput('vlnplot_1'),
               plotOutput('vlnplot_2'),
               plotOutput('feature_plot_1')
               
               
               
               
               # 
               # ```{r markerplots, fig.height=8, fig.width=15,}
               # VlnPlot(object = pbmc, features.plot = c("MS4A1", "CD79A"))
               # 
               # # you can plot raw UMI counts as well
               # VlnPlot(object = pbmc, features.plot = c("NKG7", "PF4"), use.raw = TRUE, y.log = TRUE)
               # 
               # FeaturePlot(object = pbmc, features.plot = c("MS4A1", "GNLY", "CD3E", "CD14", "FCER1A", "FCGR3A", "LYZ", "PPBP", "CD8A"), cols.use = c("grey", "blue"), reduction.use = "tsne")
      ),
      tabPanel('DoHeatmap',
               tags$h5("`DoHeatmap` generates an expression heatmap for given cells and genes. In this case, we are plotting the top 20 markers (or all markers if less than 20) for each cluster."),
               
               # ```{r clusterHeatmap, fig.height=8, fig.width=15, message=FALSE, warning=FALSE}
               # pbmc.markers %>% group_by(cluster) %>% top_n(10, avg_logFC) -> top10
               # # setting slim.col.label to TRUE will print just the cluster IDS instead of every cell name
               # DoHeatmap(object = pbmc, genes.use = top10$gene, slim.col.label = TRUE, remove.key = TRUE)
               # ```
               plotOutput('doheatmap')
               
               ),
      tabPanel('Assign Cells',
               tags$h4("### Assigning cell type identity to clusters"),
               
               #tags$h5("Fortunately in the case of this dataset, we can use canonical markers to easily match the unbiased clustering to known cell types:"),
               uiOutput('cluster_0_text'),
               uiOutput('cluster_1_text'),
               uiOutput('cluster_2_text'),
               uiOutput('cluster_3_text'),
               uiOutput('cluster_4_text'),
               uiOutput('cluster_5_text'),
               uiOutput('cluster_6_text'),
               uiOutput('cluster_7_text'),
               
               
               #   Cluster ID | Markers       | Cell Type
               # -----------|---------------|----------
               #   0          | IL7R          | CD4 T cells
               # 1          | CD14, LYZ     | CD14+ Monocytes
               # 2          | MS4A1         | B cells
               # 3          | CD8A          | CD8 T cells
               # 4          | FCGR3A, MS4A7 | FCGR3A+ Monocytes
               # 5          | GNLY, NKG7    | NK cells
               # 6          | FCER1A, CST3  | Dendritic Cells
               # 7          | PPBP          | Megakaryocytes
               # 
               # 
               # ```{r labelplot, fig.height=5, fig.width=9, warning = FALSE}
               # current.cluster.ids <- c(0, 1, 2, 3, 4, 5, 6, 7)
               # new.cluster.ids <- c("CD4 T cells", "CD14+ Monocytes", "B cells", "CD8 T cells", "FCGR3A+ Monocytes", "NK cells", "Dendritic cells", "Megakaryocytes")
               # pbmc@ident <- plyr::mapvalues(x = pbmc@ident, from = current.cluster.ids, to = new.cluster.ids)
               # TSNEPlot(object = pbmc, do.label = TRUE, pt.size = 0.5)
               # ```
               # 
               plotOutput('assign_clusters'),
               
               tags$h5("### Further subdivisions within cell types"),
               # 
               tags$h4("If you perturb some of our parameter choices above (for example, setting  `resolution=0.8` or changing the number of PCs), you might see the CD4 T cells subdivide into two groups. You can explore this subdivision to find markers separating the two T cell subsets. However, before reclustering (which will overwrite `object@ident`), we can stash our renamed identities to be easily recovered later.")
               # 
               # ```{r investigate_split, fig.width=15}
               # # First lets stash our identities for later
               # pbmc <- StashIdent(object = pbmc, save.name = "ClusterNames_0.6")
               # 
               # # Note that if you set save.snn=T above, you don't need to recalculate the SNN, and can simply put: 
               # # pbmc <- FindClusters(pbmc,resolution = 0.8)
               # pbmc <- FindClusters(object = pbmc, reduction.type = "pca", dims.use = 1:10, resolution = 0.8, print.output = FALSE)
               # 
               # # Demonstration of how to plot two tSNE plots side by side, and how to color points based on different criteria
               # plot1 <- TSNEPlot(object = pbmc, do.return = TRUE, no.legend = TRUE, do.label = TRUE)
               # plot2 <- TSNEPlot(object = pbmc, do.return = TRUE, group.by = "ClusterNames_0.6", no.legend = TRUE, do.label = TRUE)
               # plot_grid(plot1, plot2)
               # 
               # # Find discriminating markers
               # tcell.markers <- FindMarkers(object = pbmc, ident.1 = 0, ident.2 = 1)
               # 
               # # Most of the markers tend to be expressed in C1 (i.e. S100A4). However, we can see that CCR7 is upregulated in 
               # # C0, strongly indicating that we can differentiate memory from naive CD4 cells.
               # # cols.use demarcates the color palette from low to high expression
               # FeaturePlot(object = pbmc, features.plot = c("S100A4", "CCR7"), cols.use = c("green", "blue"))
               # ```
               # 
               # The memory/naive split is bit weak, and we would probably benefit from looking at more cells to see if this becomes more convincing. In the meantime, we can restore our old cluster identities for downstream processing.
               # 
               # ```{r restore}
               # pbmc <- SetAllIdent(object = pbmc, id = "ClusterNames_0.6")
               # saveRDS(pbmc, file = paste0(output_dir,output_name'_Seurat_final_output.rds'))
               # ```
               # 
               )
      
               ))  
    )
    
  )

))
